charger/mcc.go

```diff
@@ -0,0 +1,350 @@
+		return api.StatusF, nil
+	case 4, 6:
+		// 4: Paused			StatusD
+		// 6: Finished		StatusD
+		return api.StatusD, nil
+	case 5:
+		// 5: Active			StatusC
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("properties unknown result: %d", chargeState)
+	}
+}
+
+// Enabled implements the Charger.Enabled interface
+func (mcc *MobileConnect) Enabled() (bool, error) {
+	// Check if the car is connected and Paused, Active, or Finished
+	b, err := mcc.getValue(mcc.apiURL(apiChargeState))
+	if err != nil {
+		return false, err
+	}
+
+	// return value is returned in the format 0\n
+	chargeState, err := strconv.ParseInt(strings.Trim(string(b), "\n"), 10, 8)
+	if err != nil {
+		return false, err
+	}
+
+	if chargeState >= 4 && chargeState <= 6 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// Enable implements the Charger.Enable interface
+func (mcc *MobileConnect) Enable(enable bool) error {
+	// As we don't know of the API to disable charging this for now always returns an error
+	return nil
+}
+
+// MaxCurrent implements the Charger.MaxCurrent interface
+func (mcc *MobileConnect) MaxCurrent(current int64) error {
+	// The device doesn't return an error if we set a value greater than the
+	// current allowed max or smaller than the allowed min
+	// instead it will simply set it to max or min and return "OK" anyway
+	// Since the API here works differently, we fetch the limits
+	// and then return an error if the value is outside of the limits or
+	// otherwise set the new value
+	if mcc.cableInformation.MaxValue == 0 {
+		if err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentCableInformation), &mcc.cableInformation); err != nil {
+			return err
+		}
+	}
+
+	if current < mcc.cableInformation.MinValue {
+		return fmt.Errorf("value is lower than the allowed minimum value %d", mcc.cableInformation.MinValue)
+	}
+
+	if current > mcc.cableInformation.MaxValue {
+		return fmt.Errorf("value is higher than the allowed maximum value %d", mcc.cableInformation.MaxValue)
+	}
+
+	url := fmt.Sprintf("%s%d", mcc.apiURL(apiSetCurrentLimit), current)
+
+	req, err := mcc.request(http.MethodPut, url)
+	if err != nil {
+		return err
+	}
+
+	b, err := mcc.Request(req)
+	if err != nil {
+		return err
+	}
+
+	// return value is returned in the format "OK"\n
+	s := strings.Trim(string(b), "\n")
+	s = strings.Trim(s, "\"")
+
+	if s != "OK" {
+		return fmt.Errorf("Call returned an unexpected error")
+	}
+
+	return nil
+}
+
+// CurrentPower implements the Meter interface.
+func (mcc *MobileConnect) CurrentPower() (float64, error) {
+	var energy MCCEnergy
+	err := mcc.getEscapedJSON(mcc.apiURL(apiEnergy), &energy)
+
+	return energy.L1.Power + energy.L2.Power + energy.L3.Power, err
+}
+
+// ChargedEnergy implements the ChargeRater interface.
+func (mcc *MobileConnect) ChargedEnergy() (float64, error) {
+	var currentSession MCCCurrentSession
+	if err := mcc.getEscapedJSON(mcc.apiURL(apiCurrentSession), &currentSession); err != nil {
```

Comment:
Habe gerade herausgefunden, dass wenn der Charger gestartet wurde und damit noch keine Session gestartet wurde und das Auto auch nicht verbunden wurde, dann kommt kein JSON und auch kein Escaped JSON, sondern einfach nur `""` zurück. Der Code würde dann momentan `0` zurück liefern aber auch einen error mit `unexpected end of JSON input`.

Gleiches dann in `ChargingTime()`. 

Sollte da das Verhalten geändert werden?

Reply:
Wie wär statusa als default? Chargingtime wäre dann ja 0. das api ist schrott...

Reply:
"schrott" ist aber noch nett formuliert 😁 

`Status()` liefert `StatusA` in dem Fall.
 
Apropo: Diese `statusa,...` gefallen mir nicht so wirklich. Könnte man das nicht etwas "schöner" machen? :)

Reply:
Hast Du eine Idee (per Mail)?